SYCL/Basic/multi_device_context.cpp

```diff
@@ -0,0 +1,67 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+
+// REQUIRES: cuda
+
+#include <stdlib.h>
+#include <sycl.hpp>
+
+sycl::event add(sycl::queue q, sycl::buffer<int> buff, int *usm,
+                sycl::event e) {
+  return q.submit([&](sycl::handler &cgh) {
+    auto acc = buff.get_access<sycl::access::mode::read_write>(cgh);
+    cgh.depends_on(e);
+    cgh.single_task([=]() { acc[0] += *usm; });
+  });
+}
+
+int main() {
+  sycl::platform plat = sycl::platform::get_platforms()[0];
+  auto devices = plat.get_devices();
+  if (devices.size() < 2) {
+    std::cout << "Need two devices for the test!" << std::endl;
+    return 0;
+  }
+
+  sycl::device dev1 = devices[0];
+  sycl::device dev2 = devices[1];
+
+  sycl::context ctx{{dev1, dev2}};
+
+  sycl::queue q1{ctx, dev1};
+  sycl::queue q2{ctx, dev2};
+
+  int a = 1;
+  int b = 2;
+  int c = 4;
+  int d = 5;
+  {
+    sycl::buffer<int> buff1(&a, 1);
+    sycl::buffer<int> buff2(&b, 1);
+
+    // test copying usm
+    int *usm1 = sycl::malloc_device<int>(1, q1);
+    int *usm2 = sycl::malloc_device<int>(1, q2);
+    sycl::event e1 = q1.memcpy(usm1, &c, 1);
+    sycl::event e2 = q2.memcpy(usm2, &d, 1);
+
+    // test combination of usm and buffers in a kernel
+    sycl::event e3 = add(q1, buff1, usm1, e1);
```

Comment:
What is the main purpose of the test? Can we use explict `q.wait()` to eliminate the need for all the events?

Reply:
To check that common operations work when using two devices within the same context, including transferring buffer data between the devices and waiting on event , associated with an operation on another device. So while we could change each event synchronization with a wait on one or both of the queues, that would reduce the scope of the test.

Reply:
In this case, would the test somewhat reliably fail if we didn't pass the events around?

In other words, would race condition manifest itself on a reasonable HW with such small workloads?

Reply:
I believe a race condition would be relatively unlikely to manifest in such a test. But I still prefer events, as a race condition is not the only possible mode of failure.

Reply:
Also using waits would not make code much more readable, as there would need to be a wait call after every operation.

Reply:
Every other, I think. Anyway, I'm fine with events.

> as a race condition is not the only possible mode of failure

For my education, what are the others here?

Reply:
I could imagine a deadlock.

Reply:
Something completely unexpected is also always an option.